### PR TITLE
MODBULKOPS-518 - Profiles API implementation

### PR DESCRIPTION
**Pull request by kjain110**

### CODE REVIEW

**Review by TarasSpashchenko — APPROVED** [review]
## Review Summary: Profile API Implementation

Thank you for your substantial contribution adding the Profile API to `mod-bulk-operations`. This PR demonstrates excellent adherence to FOLIO conventions, robust Spring idioms, and strong test-driven development. Here are the highlights and a few minor suggestions:

### Key Strengths
- **Security and Permissions:** Each new endpoint is mapped to a granular permission in the ModuleDescriptor, with enforcement handled by Okapi—excellent for FOLIO compliance.
- **Architecture:** Clean separation of concerns with new entity, repository, mappers, service, and controller logic. CQL repository integration is idiomatic for FOLIO.
- **Testing:** Comprehensive, high-quality unit and integration tests for all new logic, including error/edge cases.
- **Documentation:** OpenAPI and JSON schemas are clear and accurate, matching the endpoints and DTOs.
- **No Security or Tenant Isolation Issues:** All repository access is tenant-scoped, and no sensitive data is exposed or logged.

### Suggestions
- **Javadoc:** Consider adding Javadoc to public service methods for future maintainability.
- **Pagination:** In `getProfileSummaries`, if available, use `page.getTotalElements()` instead of `items.stream().count()` for accurate pagination.

### No Blockers
I found no blockers or critical issues. The implementation is robust, secure, and well-tested. 

---

**Post-merge:** Please ensure CI passes on the target branch, and consider running a quick regression or smoke test for the new endpoints in a staging environment.

**Continuous Improvement:** The quality of this PR sets a great example for future contributions—especially the thorough testing and permission mapping.

---

**Status:** APPROVED

Great work!

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java
Comment (TarasSpashchenko) [inline src/main/java/org/folio/bulkops/service/BulkOperationService.java:1]: [Blocker]: According to the Jira ticket MODBULKOPS-518, PUT and DELETE operations must not be allowed for locked profiles. However, the current implementation in BulkOperationService and BulkOperationController does not enforce this restriction—locked profiles can still be updated or deleted. This is a critical business rule and must be enforced at the service layer (and ideally controller as well) to prevent accidental or unauthorized modification/deletion of locked profiles.

**Suggested fix:**
- In `BulkOperationService#deleteById(UUID id)`, check if the profile is locked before deletion. If locked, throw a relevant exception (e.g., BadRequestException or a new LockedProfileException).
- In `BulkOperationService#updateProfile(UUID profileId, ProfileUpdateRequest profileUpdateRequest)`, check if the profile is locked before updating. If locked, throw a relevant exception.
- Ensure the controller maps this exception to a 400 Bad Request or 409 Conflict response.

~~~java
if (profile.isLocked()) {
    throw new BadRequestException("Cannot delete/update a locked profile");
}
~~~

Please address this to meet the acceptance criteria and prevent data integrity issues.

File: src/main/java/org/folio/bulkops/controller/BulkOperationController.java
Comment (TarasSpashchenko) [inline src/main/java/org/folio/bulkops/controller/BulkOperationController.java:1]: [Suggestion]: The Profile API endpoints (POST/PUT/DELETE) should ideally perform input validation for required fields and constraints (e.g., name not blank, entityType not null, etc.). While some validation may be handled at the DTO level, please ensure that:
- All required fields are validated (see OpenAPI/Swagger schema requirements).
- Invalid or incomplete requests return a clear 400 Bad Request with a descriptive error message.
- Consider using `@Valid` and validation annotations on DTOs and controller methods, and handle MethodArgumentNotValidException globally if not already done.

This will improve API robustness and client experience.

File: src/main/java/org/folio/bulkops/domain/entity/Profile.java
Comment (TarasSpashchenko) [inline src/main/java/org/folio/bulkops/domain/entity/Profile.java:1]: [Suggestion]: The Profile entity now uses JSONB columns for `bulkOperationRuleCollection` and `bulkOperationMarcRuleCollection` with @JdbcTypeCode(SqlTypes.JSON). Please ensure that:
- The PostgreSQL database is configured to support JSONB columns for these fields.
- There is a Liquibase or Flyway migration to update the schema accordingly (if not already present).
- Consider adding constraints (if possible) to ensure data integrity for these JSON fields.

If this is already handled, please ignore this comment.

File: src/main/java/org/folio/bulkops/mapper/ProfileMapper.java
Comment (TarasSpashchenko) [inline src/main/java/org/folio/bulkops/mapper/ProfileMapper.java:1]: [Suggestion]: The new ProfileMapper and ProfileRequestMapper classes use MapStruct for DTO/entity mapping, which is excellent for maintainability. However, in ProfileMapper, the method name `toSummmaryDTO` has a typo (should be `toSummaryDTO`). Please correct this for clarity and consistency.

~~~java
// Current
toSummmaryDTO(Profile profile);
// Suggested
toSummaryDTO(Profile profile);
~~~

Also, ensure the usage sites are updated accordingly.

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java
Comment (TarasSpashchenko) [inline src/main/java/org/folio/bulkops/service/BulkOperationService.java:1]: [Suggestion]: In `BulkOperationService#getProfileSummaries`, the totalRecords is set as `items.stream().count()`, which only reflects the number of returned items in the current page, not the total number of records matching the query in the database. For proper pagination support, `totalRecords` should reflect the total number of matching records (i.e., `page.getTotalElements()`).

~~~java
response.setTotalRecords(page.getTotalElements());
~~~

This will allow clients to properly handle pagination and display accurate result counts.

File: src/main/java/org/folio/bulkops/repository/ProfileRepository.java
Comment (TarasSpashchenko) [inline src/main/java/org/folio/bulkops/repository/ProfileRepository.java:1]: [Suggestion]: The new ProfileRepository interface is a plain JpaRepository. If you anticipate needing custom queries (e.g., for filtering by locked status, name, or entityType), consider adding methods for those now or documenting the expected usage. This is not a blocker, but planning ahead can help avoid future refactoring.

File: src/main/resources/swagger.api/bulk-operations.yaml
Comment (TarasSpashchenko) [inline src/main/resources/swagger.api/bulk-operations.yaml:1]: [Suggestion]: The OpenAPI schema and swagger.yaml for the new Profile endpoints are well-structured and provide the necessary details. However, please ensure that the error responses for locked profile operations (PUT/DELETE) are documented to return a 400 or 409 with a meaningful error message, as per business rules. This will help front-end/UI teams and API consumers handle these cases gracefully.

File: src/test/java/org/folio/bulkops/controller/BulkOperationControllerTest.java
Comment (TarasSpashchenko) [inline src/test/java/org/folio/bulkops/controller/BulkOperationControllerTest.java:1]: [Suggestion]: The test coverage for the new Profile API is very good, with both positive and negative cases. Please consider adding a test for the locked profile scenario (attempting to update or delete a locked profile should return the expected error). This will ensure the business rule is enforced and prevent regressions.

**Review by TarasSpashchenko — CHANGES_REQUESTED** [review]
## JReview: Comprehensive Review for MODBULKOPS-518 (Profiles API implementation)

### Summary
Thank you for the substantial contribution implementing CRUD operations for bulk operation profiles in `mod-bulk-operations`! This is a significant step forward for the bulk edit platform and aligns well with the requirements from [MODBULKOPS-518](https://folio-org.atlassian.net/browse/MODBULKOPS-518). The implementation is generally robust, with thorough test coverage, clear API documentation, and good use of modern Spring idioms and FOLIO conventions.

However, there is a **blocker** that must be addressed before merging:

---

### [Blocker]
- **Locked Profile Enforcement:**
  - The business rule that PUT and DELETE operations must not be allowed for locked profiles is not enforced in the current implementation. This is a critical requirement from the Jira issue and must be handled at the service (and ideally controller) layer to prevent accidental or unauthorized modification/deletion of locked profiles. Please add checks to both `updateProfile` and `deleteById` to reject operations on locked profiles, and ensure the controller returns a suitable error response (400 or 409).

---

### Additional Suggestions
- **Input Validation:**
  - Ensure all required fields for profiles are validated using `@Valid` and relevant annotations. Handle validation exceptions globally for a better API experience.

- **Pagination Total Records:**
  - In `getProfileSummaries`, use `page.getTotalElements()` for `totalRecords` instead of counting the current page. This is important for correct pagination.

- **MapStruct Typo:**
  - Fix the typo in `ProfileMapper` method `toSummmaryDTO` → `toSummaryDTO`.

- **JSONB Schema Migration:**
  - Ensure the DB migration for new JSONB columns is present and tested.

- **Repository Extensibility:**
  - Consider if custom queries for `ProfileRepository` are needed in the future.

- **API Documentation:**
  - Document error scenarios for locked profile operations in the OpenAPI spec.

- **Test Coverage:**
  - Add tests for locked profile enforcement (controller and service layer).

---

### Security & FOLIO Compliance
- Permissions for the new endpoints are correctly defined in the ModuleDescriptor and enforced via Okapi.
- No sensitive information is exposed or logged.
- Tenant isolation is maintained via standard FOLIO patterns.

---

### Next Steps
- Please address the blocker and consider the suggestions above.
- After changes, re-run the CI pipeline and ensure all tests pass.
- For this new bulk operation logic, consider running targeted integration tests in a staging environment.

---

Thank you for your hard work and attention to detail. Please reach out if you have any questions or need clarification on any of the review points!

---

**Status:** REQUEST_CHANGES

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java
Comment (NikitaSedyx) [reply src/main/java/org/folio/bulkops/service/BulkOperationService.java:1]: good point, lock permission should be checked as well

File: src/main/java/org/folio/bulkops/controller/BulkOperationController.java [hunk 1]
```diff
@@ -256,4 +260,30 @@ public ResponseEntity<List<String>> getListUsedTenants(UUID operationId) {
   public ResponseEntity<BulkOperationDto> triggerBulkEditByQuery(UUID xOkapiUserId, QueryRequest queryRequest) {
     return new ResponseEntity<>(bulkOperationMapper.mapToDto(bulkOperationService.triggerByQuery(xOkapiUserId, queryRequest)), HttpStatus.OK);
   }
+
+
+  @Override
+  public ResponseEntity<ProfileSummaryResultsDto> getProfiles(String query, Integer offset, Integer limit) {
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/bulkops/controller/BulkOperationController.java:266]: why not to add new controller for profiles?

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java
Comment (kjain110) [reply src/main/java/org/folio/bulkops/service/BulkOperationService.java:1]: Done

File: src/main/java/org/folio/bulkops/service/BulkOperationService.java [hunk 2]
```diff
@@ -841,4 +846,77 @@ private int getProgressMultiplier(BulkOperation operation, boolean hasAdministra
     }
     return 1;
   }
+
+  public ProfileSummaryResultsDto getProfileSummaries(String query, Integer offset, Integer limit) {
```
[the remark this answers is not in the document]
Comment (kjain110) [reply src/main/java/org/folio/bulkops/service/BulkOperationService.java]: @NikitaSedyx  done

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -298,6 +298,30 @@
...
+          "methods": [ "DELETE" ],
+          "pathPattern": "/bulk-operations/profiles/{id}",
+          "permissionsRequired": ["bulk-operations.profiles.item.delete"],
+          "modulePermissions": [ ]
+        },
+        {
+          "methods": [ "PUT" ],
+          "pathPattern": "/bulk-operations/profiles/{id}",
+          "permissionsRequired": ["bulk-operations.profiles.item.put"],
+          "modulePermissions": [ ]
```
[the remark this answers is not in the document]
Comment (NikitaSedyx) [reply descriptors/ModuleDescriptor-template.json:324]: do not see any changes, you still need section with desired permissions

File: src/main/java/org/folio/bulkops/service/UserPermissionsService.java [hunk 4]
```diff
@@ -54,8 +55,8 @@ public List<String> getPermissions() {
   private List<String> getDesiredPermissions() {
     return List.of(BULK_EDIT_INVENTORY_WRITE_PERMISSION.getValue(), BULK_EDIT_USERS_WRITE_PERMISSION.getValue(),
       USERS_ITEM_PUT.getValue(), INVENTORY_ITEMS_ITEM_PUT.getValue(), INVENTORY_STORAGE_HOLDINGS_ITEM_PUT.getValue(),
-            INVENTORY_INSTANCES_ITEM_PUT.getValue(), BULK_EDIT_INVENTORY_VIEW_PERMISSION.getValue(), BULK_EDIT_USERS_VIEW_PERMISSION.getValue(),
-            USER_ITEM_GET_PERMISSION.getValue(), INVENTORY_ITEMS_ITEM_GET_PERMISSION.getValue(),
-            INVENTORY_INSTANCES_ITEM_GET_PERMISSION.getValue(), INVENTORY_STORAGE_HOLDINGS_ITEM_GET_PERMISSION.getValue());
+      INVENTORY_INSTANCES_ITEM_PUT.getValue(), BULK_EDIT_INVENTORY_VIEW_PERMISSION.getValue(), BULK_EDIT_USERS_VIEW_PERMISSION.getValue(),
+      USER_ITEM_GET_PERMISSION.getValue(), INVENTORY_ITEMS_ITEM_GET_PERMISSION.getValue(),
+      INVENTORY_INSTANCES_ITEM_GET_PERMISSION.getValue(), INVENTORY_STORAGE_HOLDINGS_ITEM_GET_PERMISSION.getValue(), BULK_OPERATIONS_PROFILES_ITEM_LOCK.getValue());
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/bulkops/service/UserPermissionsService.java:60]: looks at https://github.com/folio-org/mod-invoice/blob/master/descriptors/ModuleDescriptor-template.json#L60 , it's an example how desired perms work on module level, `getDesiredPermissions` is util to work with a case when permissions are declared in other modules

File: src/main/java/org/folio/bulkops/controller/BulkOperationController.java (hunk 1 above)
Comment (NikitaSedyx) [reply src/main/java/org/folio/bulkops/controller/BulkOperationController.java:268]: I can not agree as it breaks SRP, @khandramai please decide

File: src/main/java/org/folio/bulkops/controller/BulkOperationController.java [hunk 5]
```diff
@@ -256,4 +262,30 @@ public ResponseEntity<List<String>> getListUsedTenants(UUID operationId) {
...
+    var response = profileService.getProfileSummaries(query, offset, limit);
+    return ResponseEntity.ok(response);
+  }
+
+
+  @Override
+  public ResponseEntity<ProfileDto> createProfile(ProfileRequest profileRequest) {
```
Comment (khandramai) [inline src/main/java/org/folio/bulkops/controller/BulkOperationController.java:275]: @kjain110  , desired permission "bulk-operations.profiles.item.lock.execute" must be explicitly checked to allow or restrict creating of profile with locked=true.

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (khandramai) [reply descriptors/ModuleDescriptor-template.json:330]: @kjain110 , yes - it will be nice to implement separate controller for profiles functionality.

File: descriptors/ModuleDescriptor-template.json [hunk 6]
```diff
@@ -298,6 +298,36 @@
...
+          "pathPattern": "/bulk-operations/profiles",
+          "permissionsRequired": ["bulk-operations.profiles.item.post"],
+          "modulePermissions": [ ]
+        },
+        {
+          "methods": [ "DELETE" ],
+          "pathPattern": "/bulk-operations/profiles/{id}",
+          "permissionsRequired": ["bulk-operations.profiles.item.delete"],
+          "permissionsDesired": [
+            "bulk-operations.profiles.item.lock.execute"
```
Comment (NikitaSedyx) [inline descriptors/ModuleDescriptor-template.json:319]: please move it to POST, as POST and PUT can lock/unlock

File: src/main/java/org/folio/bulkops/service/ProfileService.java [hunk 7]
```diff
@@ -77,18 +81,25 @@ public void deleteById(UUID id) {
...
       .orElseThrow(() -> new NotFoundException("Profile not found with ID: " + profileId));
+
     if (existing.isLocked()) {
       throw new ProfileLockedException("Cannot update a locked profile");
     }
+
+    Boolean requestedLocked = profileUpdateRequest.getLocked();
+    boolean isLockingNow = Boolean.TRUE.equals(requestedLocked) && !existing.isLocked();
+
+    if (isLockingNow) {
```
[the remark this answers is not in the document]
Comment (kjain110) [reply src/main/java/org/folio/bulkops/service/ProfileService.java:99]: But it was a requirement that if a profile is locked then we can not update it

[the remark this answers is not in the document]
Comment (NikitaSedyx) [reply src/main/java/org/folio/bulkops/service/ProfileService.java:99]: there is a requirement that user with lock permission can unlock profile, so overall user with permission can edit locked profiles + only such users can change isLocked flag state, please review one more time https://folio-org.atlassian.net/browse/UIBULKED-661

[the remark this answers is not in the document]
Comment (kjain110) [reply src/main/java/org/folio/bulkops/service/ProfileService.java]: @NikitaSedyx done

File: src/main/java/org/folio/bulkops/service/ProfileService.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    Profile entity = profileRequestMapper.toEntity(profileRequest, userId, username);
+    Profile saved = profileRepository.save(entity);
+    return profileMapper.toDto(saved);
+  }
+
+  public void deleteById(UUID id) {
+    Profile profile = profileRepository.findById(id)
+      .orElseThrow(() -> new NotFoundException("Profile not found with ID: " + id));
+
+    if (profile.isLocked()) {
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/bulkops/service/ProfileService.java:79]:
```
  if (profile.isLocked()) {
    validator.checkIfLockPermissionExists();
    // No need to catch and rethrow
  }
```

File: src/main/java/org/folio/bulkops/exception/BulkOperationsExceptionHandler.java [hunk 9]
```diff
@@ -16,4 +16,8 @@ public ResponseEntity<String> handleNotFoundException(final NotFoundException e)
   public ResponseEntity<String> handleIllegalOperationStateException(final IllegalOperationStateException e) {
     return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
   }
+  @ExceptionHandler(ProfileLockedException.class)
+  public ResponseEntity<String> handleProfileLockedException(final ProfileLockedException e) {
+    return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/bulkops/exception/BulkOperationsExceptionHandler.java:21]: let's make it 403 instead of 400

File: src/main/java/org/folio/bulkops/service/ProfileService.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      } catch (ProfileLockedException e) {
+        throw new ProfileLockedException("Cannot delete a locked profile without proper permission");
+      }
+    }
+
+    profileRepository.delete(profile);
+  }
+
+  public ProfileDto updateProfile(UUID profileId, ProfileUpdateRequest profileUpdateRequest) {
+    User user = userClient.getUserById(folioExecutionContext.getUserId().toString());
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/bulkops/service/ProfileService.java:91]: you don't need to fetch user to get their UUID, just
```
    UUID userId = folioExecutionContext.getUserId();
    String username = getUsername(userId);
```

File: src/main/java/org/folio/bulkops/service/ProfileService.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        Objects.requireNonNullElse(offset, 0),
+        Objects.requireNonNullElse(limit, Integer.MAX_VALUE)
+      )
+    );
+
+    List<org.folio.bulkops.domain.dto.ProfileSummaryDTO> items = page.stream()
+      .map(profileMapper::toSummaryDTO)
+      .toList();
+
+    var response = new org.folio.bulkops.domain.dto.ProfileSummaryResultsDto();
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/bulkops/service/ProfileService.java:55]:
```
    return new ProfileSummaryResultsDto()
      .content(items)
      .totalRecords(page.getTotalElements());
```

File: src/main/java/org/folio/bulkops/service/ProfileService.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      .map(profileMapper::toSummaryDTO)
+      .toList();
+
+    return new ProfileSummaryResultsDto()
+      .content(items)
+      .totalRecords(page.getTotalElements());
+  }
+
+  public ProfileDto createProfile(org.folio.bulkops.domain.dto.ProfileRequest profileRequest) {
+    UUID userId = folioExecutionContext.getUserId();
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/bulkops/service/ProfileService.java:60]:
```    
    UUID userId = folioExecutionContext.getUserId();
    String username = getUsername(userId);
```

**Review by obozhko-folio — APPROVED** [review]

**Review by NikitaSedyx — APPROVED** [review]
